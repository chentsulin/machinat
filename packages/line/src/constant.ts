export const CHANNEL_REQUEST_GETTER = Symbol(
  'channel_request_getter.line.machinat'
);
export const BULK_REQUEST_GETTER = Symbol('bulk_request_getter.line.machinat');

/** @internal */
export const LINE = 'line' as const;

/** @internal */
export const PATH_REPLY = 'v2/bot/message/reply';

/** @internal */
export const PATH_PUSH = 'v2/bot/message/push';

/** @internal */
export const PATH_MULTICAST = 'v2/bot/message/multicast';

/** @internal */
export const PATH_RICHMENU = 'v2/bot/richmenu';

export enum LiffContextOs {
  Ios,
  Android,
  Web,
}
