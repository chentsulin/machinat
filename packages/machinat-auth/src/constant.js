// @flow
import { namedInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const AUTH_SERVER_AUTHORIZERS_I = namedInterface<
  ServerAuthorizer<any, any>[]
>({
  name: 'ServerAuthorizersList',
  multi: true,
});

export const AUTH_MODULE_CONFIGS_I = namedInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
