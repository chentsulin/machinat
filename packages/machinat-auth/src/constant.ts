import { makeInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const SERVER_AUTHORIZERS_I = makeInterface<ServerAuthorizer<any, any>[]>(
  { name: 'ServerAuthorizersList', multi: true }
);

export const MODULE_CONFIGS_I = makeInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
