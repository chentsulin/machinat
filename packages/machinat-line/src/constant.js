// @flow
export const LINE = 'line';
export const LINE_NAMESPACE = 'Line';

export const LINE_NATIVE_TYPE = Symbol.for('machinat.line.native');

export const ENTRY_GETTER = Symbol.for('machinat.line.segment.entry_getter');

export const PATH_REPLY = 'v2/bot/message/reply';
export const PATH_PUSH = 'v2/bot/message/push';
export const PATH_MULTICAST = 'v2/bot/message/multicast';
