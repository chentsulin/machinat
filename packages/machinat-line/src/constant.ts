export const LINE = 'line';

export const CHANNEL_API_CALL_GETTER = Symbol(
  'line.segment.channel_api_call_getter'
);
export const BULK_API_CALL_GETTER = Symbol('line.segment.bulk_api_call_getter');

export const PATH_REPLY = 'v2/bot/message/reply';
export const PATH_PUSH = 'v2/bot/message/push';
export const PATH_MULTICAST = 'v2/bot/message/multicast';
export const PATH_RICHMENU = 'v2/bot/richmenu';
