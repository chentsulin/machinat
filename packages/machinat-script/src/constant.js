// @flow
export const MACHINAT_SCRIPT_TYPE = Symbol.for('machinat.component.script');

export const SCRIPT_STATE_KEY = '$$machinat:script';
