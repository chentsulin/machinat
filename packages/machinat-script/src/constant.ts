import { makeInterface } from '@machinat/core/service';
import type { MachinatScript } from './types';

export const MACHINAT_SCRIPT_TYPE = Symbol.for('machinat.component.script');

export const SCRIPT_STATE_KEY = '$$machinat:script';

export const SCRIPT_LIBS_I = makeInterface<MachinatScript<any, any, any, any>>({
  name: 'ScriptLibsList',
  multi: true,
});
