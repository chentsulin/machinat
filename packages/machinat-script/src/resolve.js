// @flow
/* eslint no-use-before-define: ["error", { "variables": false }] */
import invariant from 'invariant';
import { reduce, isElement, formatNode } from 'machinat-utility';
import type { NodeReducer } from 'machinat-utility/types';
import * as KEYWORDS from './keyword';
import { isKeyword, isScript, counter } from './utils';
import type {
  MachinatScriptNode,
  VarsMatcher,
  ScriptSegment,
  IfSegment,
  ForSegment,
  WhileSegment,
  SetVarsSegment,
  PromptSegment,
  LabelSegment,
  CallSegment,
} from './types';

const ifChildrenReducer: NodeReducer<
  IfSegment,
  { condition: VarsMatcher, stopPointCounter: () => number }
> = (segment, node, path, { condition, stopPointCounter }) => {
  invariant(
    isElement(node) &&
      (node.type === KEYWORDS.Then ||
        node.type === KEYWORDS.Else ||
        node.type === KEYWORDS.ElseIf),
    `only <[Then, ElseIf, Else]/> accepted within children of <If/>, got: ${formatNode(
      node
    )}`
  );

  const { branches, fallback } = segment;

  if (node.type === KEYWORDS.Then) {
    invariant(
      branches.length === 0,
      '<Then /> should be the first block wihtin <If />'
    );

    branches.push({
      condition,
      body: resolveSegments(
        node.props.children,
        `${path}.children`,
        stopPointCounter
      ),
    });
  } else if (node.type === KEYWORDS.ElseIf) {
    invariant(
      branches.length > 0 && !fallback,
      '<ElseIf /> should be placed between <Then /> and <Else /> blocks'
    );

    invariant(
      typeof node.props.condition === 'function',
      'prop "condition" of <ElseIf/> should be a function'
    );

    segment.branches.push({
      condition: node.props.condition,
      body: resolveSegments(
        node.props.children,
        `${path}.children`,
        stopPointCounter
      ),
    });
  } else if (node.type === KEYWORDS.Else) {
    invariant(
      branches.length > 0 && !fallback,
      fallback
        ? 'multiple <Else/> block received in <If/>'
        : 'no <Then/> block before <Else/>'
    );

    // eslint-disable-next-line no-param-reassign
    segment.fallback = resolveSegments(
      node.props.children,
      `${path}.children`,
      stopPointCounter
    );
  }

  return segment;
};

const resolveIfSegment = (
  props: Object,
  path: string,
  stopPointCounter: () => number
): IfSegment => {
  const { condition, children, key } = props;

  invariant(
    typeof condition === 'function',
    'prop "condition" of <If/> should be a function'
  );

  return reduce(
    children,
    ifChildrenReducer,
    {
      type: 'if',
      branches: [],
      fallback: undefined,
      key,
    },
    `${path}.children`,
    { condition, stopPointCounter }
  );
};

const resolveForSegment = (
  props: Object,
  path: string,
  stopPointCounter: () => number
): ForSegment => {
  const { var: varName, of: getIterable, children, key } = props;

  invariant(
    typeof getIterable === 'function',
    'prop "of" of <For/> should be a function retruns iterable'
  );

  return {
    type: 'for',
    varName,
    getIterable,
    body: resolveSegments(children, `${path}.children`, stopPointCounter),
    key,
  };
};

const resolveWhileSegment = (
  props: Object,
  path: string,
  stopPointCounter: () => number
): WhileSegment => {
  const { condition, children, key } = props;

  invariant(
    typeof condition === 'function',
    'prop "condition" of <While/> should be a function'
  );

  return {
    type: 'while',
    condition,
    body: resolveSegments(children, `${path}.children`, stopPointCounter),
    key,
  };
};

const resolveVarsSegment = ({ set: setter }: Object): SetVarsSegment => {
  invariant(typeof setter === 'function', ``);

  return {
    type: 'set_vars',
    setter,
  };
};

const resolveLabelSegment = ({ key }: Object): LabelSegment => {
  invariant(key, 'prop "key" of <Label/> should not be empty');

  return {
    type: 'label',
    key,
  };
};

const resolvePromptSegment = (
  { set: setter, key }: Object,
  stopPointCounter: () => number
): PromptSegment => {
  const count = stopPointCounter();
  return {
    type: 'prompt',
    setter,
    key: key || `prompt#${count}`,
  };
};

const resolveCallSegemnt = (
  props: Object,
  stopPointCounter: () => number
): CallSegment => {
  const { script, withVars, goto: gotoKey, key } = props;
  invariant(isScript(script), `invalid "script" prop received on <Call/>`);

  if (gotoKey) {
    invariant(
      script._keyMapping.has(gotoKey),
      `key "${gotoKey}" not found in ${script.name}`
    );
  }

  const count = stopPointCounter();
  return {
    type: 'call',
    script,
    withVars,
    gotoKey,
    key: key || `call#${count}`,
  };
};

const segmentsReducer: NodeReducer<
  ScriptSegment[],
  { stopPointCounter: () => number }
> = (segments, node, path, { stopPointCounter }) => {
  if (isElement(node)) {
    const { type } = node;
    invariant(isKeyword(type), `unexpected element: ${formatNode(node)}`);

    let segment;
    if (type === KEYWORDS.If) {
      segment = resolveIfSegment(node.props, path, stopPointCounter);
    } else if (type === KEYWORDS.For) {
      segment = resolveForSegment(node.props, path, stopPointCounter);
    } else if (type === KEYWORDS.While) {
      segment = resolveWhileSegment(node.props, path, stopPointCounter);
    } else if (type === KEYWORDS.Vars) {
      segment = resolveVarsSegment(node.props);
    } else if (type === KEYWORDS.Prompt) {
      segment = resolvePromptSegment(node.props, stopPointCounter);
    } else if (type === KEYWORDS.Label) {
      segment = resolveLabelSegment(node.props);
    } else if (type === KEYWORDS.Call) {
      segment = resolveCallSegemnt(node.props, stopPointCounter);
    } else {
      invariant(false, `unexpected keyword: ${formatNode(node)}`);
    }

    segments.push(segment);
  } else {
    invariant(
      typeof node === 'function',
      `invalid script node: ${formatNode(node)}`
    );

    segments.push({
      type: 'content',
      render: node,
    });
  }

  return segments;
};

const resolveSegments = (
  node: MachinatScriptNode,
  path: string,
  stopPointCounter: () => number
) => {
  return reduce(node, segmentsReducer, [], path, { stopPointCounter });
};

const resolve = (node: MachinatScriptNode) => {
  return resolveSegments(node, '$', counter());
};

export default resolve;
