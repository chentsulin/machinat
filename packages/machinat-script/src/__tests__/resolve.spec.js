import Machinat from 'machinat';
import { MACHINAT_SCRIPT_TYPE } from '../constant';
import {
  If,
  Then,
  ElseIf,
  Else,
  For,
  While,
  Prompt,
  Vars,
  Label,
  Call,
} from '../keyword';
import resolve from '../resolve';

const AnotherScript = {
  $$typeof: MACHINAT_SCRIPT_TYPE,
  Init: () => '(Init)',
  name: 'SomeQuestions',
  _executable: [
    { type: 'content', render: () => '...' },
    { type: 'prompt', key: 'ask' },
  ],
  _keyMapping: { foo: 3, bar: 8 },
};

it('resolve content rendering fn', () => {
  const segments = resolve(
    <>
      {() => 'foo'}
      {() => 'bar'}
      {() => 'baz'}
    </>
  );
  expect(segments).toEqual([
    { type: 'content', render: expect.any(Function) },
    { type: 'content', render: expect.any(Function) },
    { type: 'content', render: expect.any(Function) },
  ]);
  expect(segments[0].render({})).toBe('foo');
  expect(segments[1].render({})).toBe('bar');
  expect(segments[2].render({})).toBe('baz');
});

describe('resolving <If/> segment', () => {
  it('resolve ok', () => {
    const segments = resolve(
      <If condition={() => true} key="an_if">
        <Then>{() => 'foo'}</Then>
      </If>
    );
    expect(segments).toEqual([
      {
        type: 'if',
        key: 'an_if',
        branches: [
          {
            condition: expect.any(Function),
            body: [{ type: 'content', render: expect.any(Function) }],
          },
        ],
        fallback: undefined,
      },
    ]);
    expect(segments[0].branches[0].condition()).toBe(true);
    expect(segments[0].branches[0].body[0].render()).toBe('foo');
  });

  it('resolve ok with else', () => {
    const segments = resolve(
      <If condition={() => true} key="an_if">
        <Then>{() => 'foo'}</Then>
        <Else>{() => 'bar'}</Else>
      </If>
    );
    expect(segments).toEqual([
      {
        type: 'if',
        key: 'an_if',
        branches: [
          {
            condition: expect.any(Function),
            body: [{ type: 'content', render: expect.any(Function) }],
          },
        ],
        fallback: [{ type: 'content', render: expect.any(Function) }],
      },
    ]);
    expect(segments[0].fallback[0].render()).toBe('bar');
  });

  it('resolve ok with else if conditions', () => {
    const segments = resolve(
      <If condition={() => true}>
        <Then>{() => 'foo'}</Then>
        <ElseIf condition={() => false}>{() => 'bar'}</ElseIf>
        <ElseIf condition={() => true}>{() => 'baz'}</ElseIf>
        <Else>{() => 'boom boom pow'}</Else>
      </If>
    );
    expect(segments).toEqual([
      {
        type: 'if',
        branches: new Array(3).fill({
          condition: expect.any(Function),
          body: [{ type: 'content', render: expect.any(Function) }],
        }),
        fallback: [{ type: 'content', render: expect.any(Function) }],
      },
    ]);
    expect(segments[0].branches[0].condition()).toBe(true);
    expect(segments[0].branches[0].body[0].render()).toBe('foo');
    expect(segments[0].branches[1].condition()).toBe(false);
    expect(segments[0].branches[1].body[0].render()).toBe('bar');
    expect(segments[0].branches[2].condition()).toBe(true);
    expect(segments[0].branches[2].body[0].render()).toBe('baz');
    expect(segments[0].fallback[0].render()).toBe('boom boom pow');
  });

  it('resolve ok nested', () => {
    expect(
      resolve(
        <If condition={() => true}>
          <Then>
            {() => 'foo'}
            <If condition={() => true}>
              <Then>{() => 'fooo'}</Then>
            </If>
            {() => 'foooo'}
          </Then>
          <ElseIf condition={() => true}>
            {() => 'bar'}
            <If condition={() => true}>
              <Then>{() => 'baar'}</Then>
              <ElseIf condition={() => true}>{() => 'baaar'}</ElseIf>
            </If>
            {() => 'baaaar'}
          </ElseIf>
          <Else>
            {() => 'baz'}
            <If condition={() => true}>
              <Then>{() => 'baaz'}</Then>
              <Else>{() => 'baaaz'}</Else>
            </If>
            {() => 'baaaaz'}
          </Else>
        </If>
      )
    ).toMatchSnapshot();
  });

  it('resolve ok if no children blocks', () => {
    expect(resolve(<If condition={() => true} key="foo"></If>)).toEqual([
      { type: 'if', key: 'foo', branches: [] },
    ]);
  });

  it('throw if condition is not a function', () => {
    expect(() => resolve(<If></If>)).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"condition\\" of <If/> should be a function"`
    );
  });

  it('throw if non Then, ElseIf, Else block node contained', () => {
    expect(() =>
      resolve(<If condition={() => true}>hello</If>)
    ).toThrowErrorMatchingInlineSnapshot(
      `"only <[Then, ElseIf, Else]/> accepted within children of <If/>, got: \\"hello\\""`
    );
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Prompt />
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"only <[Then, ElseIf, Else]/> accepted within children of <If/>, got: <Symbol(machinat.script.keyword.prompt) />"`
    );
  });

  it('throw if multiple <Then/> received', () => {
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Then>{() => 'bar'}</Then>
          <Then>{() => 'foo'}</Then>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"<Then /> should be the first block wihtin <If />"`
    );
  });

  it('throw if no <Then/> provided', () => {
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Else>{() => 'baz'}</Else>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(`"no <Then/> block before <Else/>"`);
  });

  it('throw if multiple <Else/> received', () => {
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Then>{() => 'foo'}</Then>
          <Else>{() => 'bar'}</Else>
          <Else>{() => 'baz'}</Else>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"multiple <Else/> block received in <If/>"`
    );
  });

  it('throw if <ElseIf/> not after <Then/> and before <Else/>', () => {
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Then>{() => 'foo'}</Then>
          <Else>{() => 'baz'}</Else>
          <ElseIf condition={() => true}>{() => 'bar'}</ElseIf>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"<ElseIf /> should be placed between <Then /> and <Else /> blocks"`
    );
    expect(() =>
      resolve(
        <If condition={() => true}>
          <ElseIf condition={() => true}>{() => 'bar'}</ElseIf>
          <Then>{() => 'foo'}</Then>
          <Else>{() => 'baz'}</Else>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"<ElseIf /> should be placed between <Then /> and <Else /> blocks"`
    );
  });

  it('throw if condition of <ElseIf/> is not a function', () => {
    expect(() =>
      resolve(
        <If condition={() => true}>
          <Then>{() => 'foo'}</Then>
          <ElseIf>{() => 'bar'}</ElseIf>
        </If>
      )
    ).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"condition\\" of <ElseIf/> should be a function"`
    );
  });
});

describe('resolving <For/> segment', () => {
  it('resolve ok', () => {
    const segments = resolve(
      <For var="name" of={() => ['world', 'fool', 'magician']} key="a_for">
        {({ name }) => `hello the ${name}`}
      </For>
    );
    expect(segments).toEqual([
      {
        type: 'for',
        key: 'a_for',
        varName: 'name',
        getIterable: expect.any(Function),
        body: [{ type: 'content', render: expect.any(Function) }],
      },
    ]);
    expect(segments[0].getIterable({})).toEqual(['world', 'fool', 'magician']);
    expect(segments[0].body[0].render({ name: 'strength' })).toBe(
      'hello the strength'
    );
  });

  it('resolve ok nested', () => {
    const segments = resolve(
      <For var="name" of={() => ['wands', 'swords', 'cups']} key="father_for">
        <For var="num" of={() => [1, 2, 3, 4, 5, 6]} key="child_for">
          {({ name, num }) => `${name} ${num}`}
        </For>
      </For>
    );
    expect(segments).toEqual([
      {
        type: 'for',
        key: 'father_for',
        varName: 'name',
        getIterable: expect.any(Function),
        body: [
          {
            type: 'for',
            key: 'child_for',
            varName: 'num',
            getIterable: expect.any(Function),
            body: [{ type: 'content', render: expect.any(Function) }],
          },
        ],
      },
    ]);
    expect(segments[0].getIterable({})).toEqual(['wands', 'swords', 'cups']);
    expect(segments[0].body[0].getIterable({})).toEqual([1, 2, 3, 4, 5, 6]);
    expect(segments[0].body[0].body[0].render({ name: 'coins', num: 5 })).toBe(
      'coins 5'
    );
  });

  it('throw if no iterable getter provided', () => {
    expect(() => resolve(<For />)).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"of\\" of <For/> should be a function retruns iterable"`
    );
  });
});

describe('resolving <While/> segment', () => {
  it('resolve ok', () => {
    const segments = resolve(
      <While condition={() => true} key="a_while">
        {() => 'Gee'}
      </While>
    );
    expect(segments).toEqual([
      {
        type: 'while',
        key: 'a_while',
        condition: expect.any(Function),
        body: [{ type: 'content', render: expect.any(Function) }],
      },
    ]);
    expect(segments[0].condition({})).toBe(true);
    expect(segments[0].body[0].render({})).toBe('Gee');
  });

  it('resolve ok nested', () => {
    const segments = resolve(
      <While condition={() => true} key="father_while">
        <While condition={() => false} key="child_while">
          {() => 'Goo'}
        </While>
      </While>
    );
    expect(segments).toEqual([
      {
        type: 'while',
        key: 'father_while',
        condition: expect.any(Function),
        body: [
          {
            type: 'while',
            key: 'child_while',
            condition: expect.any(Function),
            body: [{ type: 'content', render: expect.any(Function) }],
          },
        ],
      },
    ]);
    expect(segments[0].condition({})).toBe(true);
    expect(segments[0].body[0].condition({})).toBe(false);
    expect(segments[0].body[0].body[0].render({})).toBe('Goo');
  });

  it('throw if condition prop is not a function', () => {
    expect(() =>
      resolve(<While condition="true">{() => 'foo'}</While>)
    ).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"condition\\" of <While/> should be a function"`
    );
  });
});

describe('resolving <Vars/> segment', () => {
  it('resolve ok', () => {
    const helloSetter = () => ({ hello: 'world' });
    const greetedSetter = () => ({ greeted: true });
    expect(
      resolve(
        <>
          <Vars set={helloSetter} />
          {vars => `hello ${vars.hello}`}
          <Vars set={greetedSetter} />
        </>
      )
    ).toEqual([
      { type: 'set_vars', setter: helloSetter },
      { type: 'content', render: expect.any(Function) },
      { type: 'set_vars', setter: greetedSetter },
    ]);
  });

  it('throw if set is empty', () => {
    expect(() => resolve(<Vars />)).toThrowErrorMatchingInlineSnapshot(`""`);
    expect(() =>
      resolve(<Vars set={null} />)
    ).toThrowErrorMatchingInlineSnapshot(`""`);
  });
});

describe('resolving <Prompt/> segment', () => {
  it('resolve ok', () => {
    const answerSetter = (_, { event: { text } }) => ({ answer: text });
    expect(
      resolve(
        <>
          {() => 'but why!?'}
          <Prompt set={answerSetter} key="why" />
        </>
      )
    ).toEqual([
      { type: 'content', render: expect.any(Function) },
      { type: 'prompt', setter: answerSetter, key: 'why' },
    ]);

    expect(
      resolve(
        <>
          {() => 'where r u last night?'}
          <Prompt set={answerSetter} key="where" />
          {() => 'what r u doing?'}
          <Prompt set={answerSetter} key="what" />
        </>
      )
    ).toEqual([
      { type: 'content', render: expect.any(Function) },
      { type: 'prompt', key: 'where', setter: answerSetter },
      { type: 'content', render: expect.any(Function) },
      { type: 'prompt', key: 'what', setter: answerSetter },
    ]);
  });

  it('add default key if not provided', () => {
    const answerSetter = (_, { event: { text } }) => ({ answer: text });
    expect(resolve(<Prompt set={answerSetter} />)).toEqual([
      { type: 'prompt', setter: answerSetter, key: 'prompt#0' },
    ]);

    expect(
      resolve(
        <>
          <Prompt set={answerSetter} />
          <Prompt set={answerSetter} key="foo" />
          <Prompt set={answerSetter} />
        </>
      )
    ).toEqual([
      { type: 'prompt', key: 'prompt#0', setter: answerSetter },
      { type: 'prompt', key: 'foo', setter: answerSetter },
      { type: 'prompt', key: 'prompt#2', setter: answerSetter },
    ]);
  });
});

describe('resolving <Label/> segment', () => {
  it('resolve ok', () => {
    expect(
      resolve(
        <>
          <Label key="foo" />
          {() => 'foo'}
        </>
      )
    ).toEqual([
      { type: 'label', key: 'foo' },
      { type: 'content', render: expect.any(Function) },
    ]);

    expect(
      resolve(
        <>
          <Label key="bar" />
          {() => 'bar'}
          <Label key="baz" />
          {() => 'baz'}
        </>
      )
    ).toEqual([
      { type: 'label', key: 'bar' },
      { type: 'content', render: expect.any(Function) },
      { type: 'label', key: 'baz' },
      { type: 'content', render: expect.any(Function) },
    ]);
  });

  it('throw if key is empyt', () => {
    expect(() => resolve(<Label />)).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"key\\" of <Label/> should not be empty"`
    );
    expect(() => resolve(<Label key="" />)).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"key\\" of <Label/> should not be empty"`
    );
    expect(() =>
      resolve(<Label key={null} />)
    ).toThrowErrorMatchingInlineSnapshot(
      `"prop \\"key\\" of <Label/> should not be empty"`
    );
  });
});

describe('resolving <Call/> segments', () => {
  it('resolve ok', () => {
    const segments = resolve(
      <>
        {() => 'hello'}
        <Call
          script={AnotherScript}
          withVars={() => ({ foo: 'bar' })}
          goto="greet"
          key="waiting"
        />
      </>
    );
    expect(segments).toEqual([
      { type: 'content', render: expect.any(Function) },
      {
        type: 'call',
        script: AnotherScript,
        withVars: expect.any(Function),
        gotoKey: 'greet',
        key: 'waiting',
      },
    ]);
    expect(segments[1].withVars({})).toEqual({ foo: 'bar' });
  });

  it('throw if non-script received', () => {
    expect(() =>
      resolve(<Call script={{ something: 'wrong' }} />)
    ).toThrowErrorMatchingInlineSnapshot(
      `"invalid \\"script\\" prop received on <Call/>"`
    );
  });
});

test('resolve whole script', () => {
  expect(
    resolve(
      <>
        <Label key="start" />
        {() => <b>Lorem</b>}

        <If condition={() => false} key="first">
          <Then>
            <While condition={() => true}>
              {() => <i>ipsum</i>}
              <Prompt setter={(vars, frm) => ({ ...vars, a: frm.a })} />
            </While>
          </Then>
          <ElseIf condition={() => true}>
            <For var="n" of={() => [1, 2, 3]}>
              {() => <dolor />}
              <Prompt setter={(vars, frm) => ({ ...vars, b: frm.b })} />
            </For>
          </ElseIf>
          <Else>
            {() => 'sit amet,'}
            <Call
              script={AnotherScript}
              withVars={() => ({ foo: 'bar' })}
              goto="xxx"
            />
          </Else>
        </If>

        <Vars set={vars => ({ ...vars, foo: 'bar' })} />
        <Label key="second" />
        {() => 'consectetur'}

        <For var="x" of={() => ['adipiscing', 'elit']}>
          {() => 'sed do'}
          <If condition={() => false}>
            <Then>
              {() => <eiusmod />}
              <Prompt setter={(vars, frm) => ({ ...vars, c: frm.c })} />
            </Then>
            <Else>{() => 'tempor'}</Else>
          </If>

          <While condition={() => true}>
            <Label key="three" />
            {() => <magna />}
          </While>
        </For>

        <Vars set={vars => ({ ...vars, foo: 'baz' })} />
        <Label key="four" />
        {() => <del>incididunt</del>}

        <While condition={() => false}>
          {() => 'ut labore et'}
          <If condition={() => false}>
            <Then>
              {() => <dolore />}
              <Prompt setter={(vars, frm) => ({ ...vars, d: frm.d })} />
            </Then>
            <Else>
              <For var="y" of={() => ['ex', 'ea']}>
                {() => 'aliqua'}
              </For>
            </Else>
          </If>
        </While>

        {() => <a>Ut enim</a>}
        <Call
          script={AnotherScript}
          withVars={() => ({ foo: 'baz' })}
          goto="zzz"
        />

        <Label key="end" />
        <Prompt set={() => ({ end: true })} />
        {() => 'ad minim veniam'}
      </>
    )
  ).toMatchSnapshot();
});

it('throw if invalid syntax node received', () => {
  expect(() => resolve('hello')).toThrowErrorMatchingInlineSnapshot(
    `"invalid script node: \\"hello\\""`
  );

  expect(() => resolve(<world />)).toThrowErrorMatchingInlineSnapshot(
    `"unexpected element: <world />"`
  );

  const Foo = () => {};
  expect(() => resolve(<Foo />)).toThrowErrorMatchingInlineSnapshot(
    `"unexpected element: <Foo />"`
  );

  expect(() => resolve(<Then />)).toThrowErrorMatchingInlineSnapshot(
    `"unexpected keyword: <Symbol(machinat.script.keyword.then) />"`
  );
});
