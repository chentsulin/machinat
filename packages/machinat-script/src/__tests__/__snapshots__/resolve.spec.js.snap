// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve <IF/> resolve nested <IF/> 1`] = `
Array [
  Object {
    "branches": Array [
      Object {
        "body": Array [
          Object {
            "render": [Function],
            "type": "content",
          },
          Object {
            "branches": Array [
              Object {
                "body": Array [
                  Object {
                    "render": [Function],
                    "type": "content",
                  },
                ],
                "condition": [Function],
              },
            ],
            "fallback": undefined,
            "type": "conditions",
          },
          Object {
            "render": [Function],
            "type": "content",
          },
        ],
        "condition": [Function],
      },
      Object {
        "body": Array [
          Object {
            "render": [Function],
            "type": "content",
          },
          Object {
            "branches": Array [
              Object {
                "body": Array [
                  Object {
                    "render": [Function],
                    "type": "content",
                  },
                ],
                "condition": [Function],
              },
              Object {
                "body": Array [
                  Object {
                    "render": [Function],
                    "type": "content",
                  },
                ],
                "condition": [Function],
              },
            ],
            "fallback": undefined,
            "type": "conditions",
          },
          Object {
            "render": [Function],
            "type": "content",
          },
        ],
        "condition": [Function],
      },
    ],
    "fallback": Array [
      Object {
        "render": [Function],
        "type": "content",
      },
      Object {
        "branches": Array [
          Object {
            "body": Array [
              Object {
                "render": [Function],
                "type": "content",
              },
            ],
            "condition": [Function],
          },
        ],
        "fallback": Array [
          Object {
            "render": [Function],
            "type": "content",
          },
        ],
        "type": "conditions",
      },
      Object {
        "render": [Function],
        "type": "content",
      },
    ],
    "type": "conditions",
  },
]
`;

exports[`resolve whole script 1`] = `
Array [
  Object {
    "key": "start",
    "type": "label",
  },
  Object {
    "render": [Function],
    "type": "content",
  },
  Object {
    "branches": Array [
      Object {
        "body": Array [
          Object {
            "key": "1st",
            "type": "label",
          },
          Object {
            "body": Array [
              Object {
                "render": [Function],
                "type": "content",
              },
              Object {
                "key": "ask_1",
                "setter": [Function],
                "type": "prompt",
              },
            ],
            "condition": [Function],
            "type": "while",
          },
        ],
        "condition": [Function],
      },
      Object {
        "body": Array [
          Object {
            "render": [Function],
            "type": "content",
          },
          Object {
            "branches": Array [
              Object {
                "body": Array [
                  Object {
                    "render": [Function],
                    "type": "content",
                  },
                  Object {
                    "key": "ask_2",
                    "setter": [Function],
                    "type": "prompt",
                  },
                ],
                "condition": [Function],
              },
            ],
            "fallback": Array [
              Object {
                "render": [Function],
                "type": "content",
              },
            ],
            "type": "conditions",
          },
        ],
        "condition": [Function],
      },
    ],
    "fallback": Array [
      Object {
        "render": [Function],
        "type": "content",
      },
      Object {
        "goto": "bar",
        "key": "call_1",
        "script": Object {
          "$$typeof": Symbol(machinat.component.script),
          "commands": Array [
            Object {
              "render": [Function],
              "type": "content",
            },
            Object {
              "key": "ask",
              "type": "prompt",
            },
          ],
          "entryKeysIndex": Map {
            "foo" => 3,
            "bar" => 8,
          },
          "name": "AnotherScript",
        },
        "setter": [Function],
        "type": "call",
        "withVars": [Function],
      },
    ],
    "type": "conditions",
  },
  Object {
    "setter": [Function],
    "type": "set_vars",
  },
  Object {
    "key": "2nd",
    "type": "label",
  },
  Object {
    "render": [Function],
    "type": "content",
  },
  Object {
    "setter": [Function],
    "type": "set_vars",
  },
  Object {
    "key": "3rd",
    "type": "label",
  },
  Object {
    "render": [Function],
    "type": "content",
  },
  Object {
    "body": Array [
      Object {
        "render": [Function],
        "type": "content",
      },
      Object {
        "branches": Array [
          Object {
            "body": Array [
              Object {
                "key": "4th",
                "type": "label",
              },
              Object {
                "render": [Function],
                "type": "content",
              },
              Object {
                "key": "ask_3",
                "setter": [Function],
                "type": "prompt",
              },
            ],
            "condition": [Function],
          },
        ],
        "fallback": Array [
          Object {
            "type": "return",
            "valueGetter": [Function],
          },
        ],
        "type": "conditions",
      },
    ],
    "condition": [Function],
    "type": "while",
  },
  Object {
    "render": [Function],
    "type": "content",
  },
  Object {
    "goto": "foo",
    "key": "call_2",
    "script": Object {
      "$$typeof": Symbol(machinat.component.script),
      "commands": Array [
        Object {
          "render": [Function],
          "type": "content",
        },
        Object {
          "key": "ask",
          "type": "prompt",
        },
      ],
      "entryKeysIndex": Map {
        "foo" => 3,
        "bar" => 8,
      },
      "name": "AnotherScript",
    },
    "setter": undefined,
    "type": "call",
    "withVars": [Function],
  },
  Object {
    "key": "end",
    "type": "label",
  },
  Object {
    "key": "ask_4",
    "setter": [Function],
    "type": "prompt",
  },
  Object {
    "render": [Function],
    "type": "content",
  },
]
`;
